File: java_spring/src/main/java/com/example/java_spring/service/PersonaService.java
package com.example.java_spring.service;

import com.example.java_spring.model.Persona;
import com.example.java_spring.repository.PersonaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
public class PersonaService {

    private final PersonaRepository personaRepository;

    public PersonaService(PersonaRepository personaRepository) {
        this.personaRepository = personaRepository;
    }

    @Transactional(readOnly = true)
    public List<Persona> findAll() {
        return personaRepository.findAll();
    }

    @Transactional(readOnly = true)
    public Optional<Persona> findById(Long id) {
        return personaRepository.findById(id);
    }

    @Transactional
    public Persona save(Persona persona) {
        return personaRepository.save(persona);
    }

    @Transactional
    public Optional<Persona> update(Long id, Persona persona) {
        return personaRepository.findById(id).map(existing -> {
            existing.setNombre(persona.getNombre());
            existing.setApellido(persona.getApellido());
            existing.setDocumento(persona.getDocumento());
            existing.setDireccion(persona.getDireccion());
            return personaRepository.save(existing);
        });
    }

    @Transactional
    public Optional<Persona> patch(Long id, Persona persona) {
        return personaRepository.findById(id).map(existing -> {
            if (persona.getNombre() != null) {
                existing.setNombre(persona.getNombre());
            }
            if (persona.getApellido() != null) {
                existing.setApellido(persona.getApellido());
            }
            if (persona.getDocumento() != null) {
                existing.setDocumento(persona.getDocumento());
            }
            if (persona.getDireccion() != null) {
                existing.setDireccion(persona.getDireccion());
            }
            return personaRepository.save(existing);
        });
    }

    @Transactional
    public boolean delete(Long id) {
        if (personaRepository.existsById(id)) {
            personaRepository.deleteById(id);
            return true;
        }
        return false;
    }
}

File: kotlin_spring/src/main/kotlin/com/example/kotlin_spring/service/PersonaService.kt
package com.example.kotlin_spring.service

import com.example.kotlin_spring.model.Persona
import com.example.kotlin_spring.repository.PersonaRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class PersonaService(private val personaRepository: PersonaRepository) {

    fun findAll(): List<Persona> = personaRepository.findAll()

    fun findById(id: Long): Persona? = personaRepository.findById(id).orElse(null)

    @Transactional
    fun save(persona: Persona): Persona = personaRepository.save(persona)

    @Transactional
    fun update(id: Long, persona: Persona): Persona? {
        return if (personaRepository.existsById(id)) {
            persona.copy(id = id).let { personaRepository.save(it) }
        } else null
    }

    @Transactional
    fun patch(id: Long, fields: Map<String, Any>): Persona? {
        val existing = personaRepository.findById(id).orElse(null) ?: return null
        val updated = existing.copy(
            nombre = fields["nombre"] as? String ?: existing.nombre,
            apellido = fields["apellido"] as? String ?: existing.apellido,
            documento = fields["documento"] as? String ?: existing.documento,
            direccion = fields["direccion"] as? String ?: existing.direccion
        )
        return personaRepository.save(updated)
    }

    @Transactional
    fun delete(id: Long): Boolean {
        return if (personaRepository.existsById(id)) {
            personaRepository.deleteById(id)
            true
        } else false
    }
}